### PR TITLE
[Fix - Chunking] Safely cut strings in middle of emojis

### PR DESCRIPTION
**Pull request by philipperolet**
## Description

When chunking, we sometimes need to cut strings. If we cut in the middle of a 2-char (or more) unicode character, it falls apart. Cf this [discussion](https://dust4ai.slack.com/archives/C05F84CFP0E/p1706034315114889?thread_ts=1706009720.515859&cid=C05F84CFP0E) => this PR fixes by providing a `safeSubstring` method

## Risk
The function discards the cut emoji => a better version would ensure safeSubstring(0,128) + safeSubstring(128) = all the string even if

Scope is limited here, so risk is not high

## Notes
A later version coud directly change the substring prototype?

### CODE REVIEW

File: types/src/shared/utils/string_utils.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * Read more:
+ * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#utf-16_characters_unicode_code_points_and_grapheme_clusters
+ */
+export function safeSubstring(
+  str: string,
+  start: number,
+  end?: number
+): string {
+  while (isTrailingLoneSurrogate(str.charCodeAt(start))) start++;
```
Comment (flvndvd) [inline types/src/shared/utils/string_utils.ts:18]:
```suggestion
  while (isTrailingLoneSurrogate(str.charCodeAt(start))) {
    start++;
  }
```

File: types/src/shared/utils/string_utils.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  if (end === undefined) {
+    return str.substring(start);
+  }
+  while (isLeadingLoneSurrogate(str.charCodeAt(end - 1))) end--;
```
Comment (flvndvd) [inline types/src/shared/utils/string_utils.ts:22]:
```suggestion
  while (isLeadingLoneSurrogate(str.charCodeAt(end - 1))) {
    end--;
  }
```

**Review by flvndvd — APPROVED** [review]

**Comment by philipperolet** [comment]
test code
```const fullTitle =
  "Les croissants demain c'est pour moi, je ramène même les petits tigrés, un mi-cuit au chocolat et des paris brests mes Loulous 🩵 :innocent: ";

async function main() {
  await reproTokenize(fullTitle);
  await reproTokenize(safeSubstring(fullTitle, 0, 128));
  await reproTokenize(fullTitle.substring(0, 128));
}
main();```

**Comment by philipperolet** [comment]
output:
 ```
Tokenizing  Les croissants demain c'est pour moi, je ramène même les petits tigrés, un mi-cuit au chocolat et des paris brests mes Loulous 🩵 :innocent: 
Success ! Les croissants demain c'est pour moi, je ramène même les petits tigrés, un mi-cuit au chocolat et des paris brests mes Loulous ��� :innocent: 
Tokenizing  Les croissants demain c'est pour moi, je ramène même les petits tigrés, un mi-cuit au chocolat et des paris brests mes Loulous 
Success ! Les croissants demain c'est pour moi, je ramène même les petits tigrés, un mi-cuit au chocolat et des paris brests mes Loulous 
Tokenizing  Les croissants demain c'est pour moi, je ramène même les petits tigrés, un mi-cuit au chocolat et des paris brests mes Loulous �
node:internal/deps/undici/undici:6466
        throw new TypeError("Body is unusable");
              ^

TypeError: Body is unusable
    at specConsumeBody (node:internal/deps/undici/undici:6466:15)
    at Response.text (node:internal/deps/undici/undici:6366:18)
    at CoreAPI2._callee37$ (/home/filou/sources/dust/types/dist/types.cjs.development.js:2799:29)
    at tryCatch (/home/filou/sources/dust/types/dist/types.cjs.development.js:708:16)
    at Generator.<anonymous> (/home/filou/sources/dust/types/dist/types.cjs.development.js:796:17)
    at Generator.throw (/home/filou/sources/dust/types/dist/types.cjs.development.js:737:21)
    at asyncGeneratorStep (/home/filou/sources/dust/types/dist/types.cjs.development.js:974:23)
    at _throw (/home/filou/sources/dust/types/dist/types.cjs.development.js:996:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)

Node.js v18.15.0
```

**Review by spolu — COMMENTED** [review]
🙏🔥💯

**Pull request merged**